### PR TITLE
reside-103: Input handling

### PR DESCRIPTION
**Pull request by richfitz**
This PR implements proper query and path component input handling, alongside body handling (binary and json).  There is quite a lot extra that is *not* done yet here.  Sorry about the size - codecov suggests a net increase of ~200 lines of Real Code, which is still quite a lot.  Most of the lines here are docs and tests though.

One complication that might not be obvious is that I've had to bypass all of plumber's built in "filters" as they make a mess of the body, and they mush the query parameters in with the path parameters.  Another option would be to leave the query filter here, and grab the appropriate arguments from the args list that it creates as we already ensure that argument names are unique across all input types.

Things are are not implemented yet that I would like to push out to future work (as this is quite large)

* One disadvantage of the current approach is that typed dynamic routes are pretty clunky - one *must* use the plumber version and if a typemismatch occurs we get a 404 not a 400.
* State not yet handled, but will be done the same way (perhaps both per-endpoint and per-router)
* No input from headers
* No input from cookies
* No easy way of interrogating an endpoint as to its properties (needed for swagger support)
* Allow duplicated query parameters (`/path?a=1&a=2`), though this required a way of describing of "vector of numbers" and is a bit subtle
* Allow endpoints to accept more than one type of input (e.g., `text/csv` and `application/json` passing *either* through)
* Only allow post to be provided for endpoints that allow it (GET should not for example)
* Variadic queries (perhaps on presence of `...` in function) though no obvious way of describing types
* Currently any target function that uses `...` will fail I believe
* No support for form input, which requires more work because we'll need to accept more than one element most of the time
* No support for destructuring a json body, for example spreading out the top level keys across different arguments (even though this is a very common likely use case)
* No support for renaming an argument (mapping query parameter x to function argument y)

### CODE REVIEW

File: R/pkgapi.R
Comment (r-ash) [inline R/pkgapi.R]: Mistakenly removed new line?

File: R/endpoint.R
Comment (r-ash) [inline R/endpoint.R:119]: Can we add a note to the doc along the lines of `This is what gets called by plumber when an endpoint receives a request`?

File: R/utils.R
Comment (r-ash) [inline R/utils.R]: I think a name which covers what this does would make the code which calls this easier to read. Like `get_path_parameters` or something or `read_path_parameters` if you want to avoid get.

File: R/input.R
Comment (r-ash) [inline R/input.R]: Something other than `data` here would make it easier to read `parameters <- get_path_parameters(path)`

File: R/input.R
Comment (r-ash) [inline R/input.R]: Can re-use result of `parse_plumber_path(path)` from above

File: R/input.R
Comment (r-ash) [inline R/input.R]: I'm getting pretty confused around here. There is just lots of moving parts I'm struggling to keep track of in my head. Lots of closures being passed around and these arguments are all lists with various properties which it is hard to keep track of without types or documentation.

We talked about this in person moving to using some more objects here to hopefully tidy up some of the tricky bits.

**Review by r-ash — CHANGES_REQUESTED** [review]

File: R/utils.R
Comment (richfitz) [reply R/utils.R]: I've gone with `parse_path_parameters` for this - better?

File: R/input.R
Comment (richfitz) [reply R/input.R]: ok, as discussed, this is now overhauled and considerably better for your comments

File: R/endpoint.R
Comment (r-ash) [inline R/endpoint.R]: Should documentation on params `input_query` and `input_body` be merged into a section for `@param ...`?

File: R/input.R
Comment (r-ash) [inline R/input.R]: Several TODO's still here - do you want to address these or break into separate tickets before merging?

File: R/returning.R
Comment (r-ash) [inline R/returning.R:49]: Not directly related to the changes here but it looks like the docs about `root` aren't correct. If called with `root = NULL` then this immediately throws an error from call to `schema_root`

**Review by r-ash — APPROVED** [review]
Looks great, couple of small questions. But I don't think blocks merging.

Also noticed with this updated impl it looks like passing multiple params in query string no longer working. So e.g. in the example package in #3 I'm expecting that if you send a GET to `/book/<title>?details=author&details=location` you will get information for both. (I believe this was working previously)

Trying now sticking a debug as early as the `pkgapi_filter_query_string` and calling with different params `details=author&something=thing` it looks like only the first param is coming through in the request `QUERY_STRING`. Any idea what is going on there?

File: R/returning.R
Comment (richfitz) [reply R/returning.R:49]: Updated and moved to RESIDE-121

File: R/input.R
Comment (richfitz) [reply R/input.R]: Dealt with this now

**Pull request merged**
